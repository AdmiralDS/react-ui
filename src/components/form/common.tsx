import styled from 'styled-components';
import { FORM_FIELD_DATA_CONTAINER_ATTRIBUTE } from '#src/components/common/utils/splitDataAttributes';

const Desc = styled.div`
  font-family: 'VTB Group UI';
  font-size: 16px;
  line-height: 24px;
`;

export const DataAttributesDescription = () => (
  <Desc>
    Компоненту можно прокидывать дата аттрибуты вида [{FORM_FIELD_DATA_CONTAINER_ATTRIBUTE}-...]. Этот аттрибут попадет
    только на контейнер самого компонента.
    <br /> Например: data-container-id="fieldIdOne" - контейнер компонента получит аттрибут
    data-container-id="fieldIdOne", на нативный input этот аттрибут прокинут не будет.
  </Desc>
);
