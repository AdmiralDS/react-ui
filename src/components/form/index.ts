export * from './InputField';
export * from './CheckboxField';
export * from './DateField';
export * from './SelectField';
export * from './FieldSet';
export * from './SuggestField';
export * from './TextField';
export * from './MultiSelectField';
export * from './SliderInputField';
export * from './SliderRangeField';
export * from './TimeField';
export * from './NumberInputField';
export * from './SearchSelectField';
export * from './EditModeField';
export * from './PhoneInputField';
