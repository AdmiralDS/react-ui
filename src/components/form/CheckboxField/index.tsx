import { forwardRef, useMemo } from 'react';
import styled from 'styled-components';
import { passFormFieldContainerDataAttributes } from '#src/components/common/utils/splitDataAttributes';
import { uid } from '#src/components/common/uid';
import { keyboardKey } from '../../common/keyboardKey';
import type { CheckboxComponentDimension } from '#src/components/form/CheckboxField/style';
import {
  CheckboxComponentHint,
  CheckboxComponentLabel,
  CheckboxComponentWrapper,
} from '#src/components/form/CheckboxField/style';
import { Checkbox } from '#src/components/Checkbox';

export interface CheckboxFieldProps extends React.InputHTMLAttributes<HTMLInputElement> {
  /** Текст будет виден ниже компонента */
  extraText?: React.ReactNode;
  /** Размер компонента */
  dimension?: CheckboxComponentDimension;
  /** Состояние частичного выбора */
  indeterminate?: boolean;
  /** Состояние ошибки */
  error?: boolean;
  /** Состояние hover */
  hovered?: boolean;
}

const StyledChekbox = styled(Checkbox)`
  margin-top: 2px;
`;

export const CheckboxField = forwardRef<HTMLInputElement, CheckboxFieldProps>(
  (
    {
      children,
      disabled = false,
      readOnly = false,
      hovered,
      dimension = 'm',
      error,
      indeterminate = false,
      extraText,
      className,
      id: passedId,
      name,
      style,
      ...props
    },
    ref,
  ) => {
    //id не должно меняться на каждый ререндер если его не назначили
    const id = useMemo(() => passedId || uid(), [passedId]);

    const fieldContainerProps = {
      'data-field-id': id,
      'data-field-name': name,
    } as Record<string, string | number>;

    passFormFieldContainerDataAttributes(props, fieldContainerProps);

    const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
      if (readOnly) {
        const code = keyboardKey.getCode(e);
        if (code === keyboardKey[' ']) {
          e.preventDefault();
        }
      }

      props.onKeyDown?.(e);
    };

    return (
      <CheckboxComponentWrapper
        className={className}
        $dimension={dimension}
        disabled={disabled}
        readOnly={readOnly}
        style={style}
        {...fieldContainerProps}
        onClick={(e) => {
          // запрещает всплытие двойного клика
          // https://css-tricks.com/when-a-click-is-not-just-a-click/#aa-checking-checkboxes
          if ((e.target as HTMLElement).tagName !== 'INPUT') {
            e.stopPropagation();
          }
        }}
      >
        <StyledChekbox
          {...props}
          ref={ref}
          disabled={disabled}
          readOnly={readOnly}
          type="checkbox"
          indeterminate={indeterminate}
          onKeyDown={handleKeyDown}
          data-hovered={hovered}
          hovered={hovered}
          dimension={dimension}
          id={id}
          name={name}
          error={error}
        />
        {children && (
          <CheckboxComponentLabel $dimension={dimension} disabled={disabled}>
            {children}
            {extraText && (
              <CheckboxComponentHint disabled={disabled} $dimension={dimension}>
                {extraText}
              </CheckboxComponentHint>
            )}
          </CheckboxComponentLabel>
        )}
      </CheckboxComponentWrapper>
    );
  },
);

CheckboxField.displayName = 'CheckboxField';
