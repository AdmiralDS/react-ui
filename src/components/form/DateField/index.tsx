import type { FieldOwnProps } from '#src/components/Field';
import { Field } from '#src/components/Field';
import type { DateInputProps } from '#src/components/input/DateInput';
import { DateInput } from '#src/components/input/DateInput';
import * as React from 'react';
import { uid } from '#src/components/common/uid';
import {
  passFormFieldContainerDataAttributes,
  passFormFieldDataAttributes,
} from '#src/components/common/utils/splitDataAttributes';

export interface DateFieldProps extends DateInputProps, Omit<FieldOwnProps, 'inputRef'> {}

export const DateField = React.forwardRef<HTMLInputElement, DateFieldProps>((props, ref) => {
  const {
    className,
    displayInline,
    status,
    required,
    extraText,
    label,
    id = uid(),
    name,
    disabled,
    skeleton,
    ...restProps
  } = props;

  const fieldContainerProps = {
    className,
    extraText,
    status,
    required,
    label,
    id,
    skeleton,
    'data-field-id': id,
    'data-field-name': name,
    displayInline,
    disabled,
  } as Record<string, any>;

  passFormFieldDataAttributes(restProps, fieldContainerProps);
  passFormFieldContainerDataAttributes(restProps, fieldContainerProps);

  const inputProps = { ref, id, name, 'aria-required': required, status, disabled, skeleton, ...restProps };
  return (
    <Field {...fieldContainerProps}>
      <DateInput {...inputProps} />
    </Field>
  );
});

DateField.displayName = 'DateField';
