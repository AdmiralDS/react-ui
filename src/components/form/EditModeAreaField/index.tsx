import { forwardRef, useRef } from 'react';

import { uid } from '#src/components/common/uid';
import { refSetter } from '#src/components/common/utils/refSetter';
import type { FieldOwnProps } from '#src/components/Field';
import { Field } from '#src/components/Field';
import type { EditModeAreaProps } from '#src/components/input';
import { EditModeArea } from '#src/components/input';
import {
  passFormFieldContainerDataAttributes,
  passFormFieldDataAttributes,
} from '#src/components/common/utils/splitDataAttributes';

export interface EditModeAreaFieldProps extends EditModeAreaProps, Omit<FieldOwnProps, 'inputRef'> {}

export const EditModeAreaField = forwardRef<HTMLTextAreaElement, EditModeAreaFieldProps>((props, ref) => {
  const inputRef = useRef<HTMLTextAreaElement>(null);
  const {
    className,
    maxLength,
    displayInline,
    status,
    required,
    extraText,
    label,
    id = uid(),
    disabled,
    displayCharacterCounter,
    ...restProps
  } = props;
  const fieldContainerProps = {
    className,
    extraText,
    status,
    required,
    label,
    id,
    displayInline,
    disabled,
    maxLength,
    inputRef,
    displayCharacterCounter,
    'data-field-id': id,
    'data-field-name': restProps.name,
  } as Record<string, any>;

  passFormFieldDataAttributes(restProps, fieldContainerProps);
  passFormFieldContainerDataAttributes(restProps, fieldContainerProps);

  const inputProps = {
    ref: refSetter(ref, inputRef),
    id,
    'aria-required': required,
    status,
    disabled,
    maxLength,
    ...restProps,
  };

  return (
    <Field {...fieldContainerProps}>
      <EditModeArea {...inputProps} />
    </Field>
  );
});

EditModeAreaField.displayName = 'EditModeAreaField';
