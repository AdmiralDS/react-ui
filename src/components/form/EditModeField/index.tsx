import { uid } from '#src/components/common/uid';
import { refSetter } from '#src/components/common/utils/refSetter';
import type { FieldOwnProps } from '#src/components/Field';
import { Field } from '#src/components/Field';
import type { EditModeProps } from '#src/components/input';
import { EditMode } from '#src/components/input';
import * as React from 'react';
import {
  passFormFieldContainerDataAttributes,
  passFormFieldDataAttributes,
} from '#src/components/common/utils/splitDataAttributes';

export interface EditModeFieldProps extends EditModeProps, Omit<FieldOwnProps, 'inputRef'> {}

export const EditModeField = React.forwardRef<HTMLInputElement, EditModeFieldProps>((props, ref) => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  const {
    className,
    maxLength,
    displayInline,
    status,
    required,
    extraText,
    label,
    id = uid(),
    disabled,
    displayCharacterCounter,
    ...restProps
  } = props;
  const fieldContainerProps = {
    className,
    extraText,
    status,
    required,
    label,
    id,
    displayInline,
    disabled,
    maxLength,
    inputRef,
    displayCharacterCounter,
    'data-field-id': id,
    'data-field-name': restProps.name,
  } as Record<string, any>;

  passFormFieldDataAttributes(restProps, fieldContainerProps);
  passFormFieldContainerDataAttributes(restProps, fieldContainerProps);

  const inputProps = {
    ref: refSetter(ref, inputRef),
    id,
    'aria-required': required,
    status,
    disabled,
    maxLength,
    ...restProps,
  };

  return (
    <Field {...fieldContainerProps}>
      <EditMode {...inputProps} />
    </Field>
  );
});

EditModeField.displayName = 'EditModeField';
