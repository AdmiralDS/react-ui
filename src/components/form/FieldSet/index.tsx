import type { FC, FieldsetHTMLAttributes, ReactNode } from 'react';
import styled from 'styled-components';

import { ExtraTextContainer } from '#src/components/Field';
import type { InputStatus } from '#src/components/input';
import { typography } from '#src/components/Typography';

export const ALL_FIELDSET_DIMENSIONS_VALUES = ['m', 's'] as const;

/** Размеры компонента относительно начального */
export type FieldSetDimension = (typeof ALL_FIELDSET_DIMENSIONS_VALUES)[number];

const Legend = styled.legend<{ $dimension?: FieldSetDimension }>`
  ${(props) => (props.$dimension === 's' ? typography['Body/Body 2 Short'] : typography['Body/Body 1 Short'])}
  color: var(--admiral-color-Neutral_Neutral90, ${(p) => p.theme.color['Neutral/Neutral 90']});
  fieldset:disabled & {
    color: var(--admiral-color-Neutral_Neutral30, ${(p) => p.theme.color['Neutral/Neutral 30']});
    cursor: default;
  }

  [aria-required='true'] &:after {
    content: ' *';
    color: var(--admiral-color-Error_Error60Main, ${(p) => p.theme.color['Error/Error 60 Main']});
  }

  [data-status='error'] & {
    color: var(--admiral-color-Error_Error60Main, ${(p) => p.theme.color['Error/Error 60 Main']});
  }
`;

const FieldSetExtrasContainer = styled.div`
  flex: 1 1 100%;
  display: flex;
  margin: 0;
`;

const FieldSetContainer = styled.fieldset<{
  $dimension?: FieldSetDimension;
  $flexDirection?: 'column' | 'row';
}>`
  border: none;
  padding: 0;

  display: flex;
  flex-wrap: wrap;
  flex-direction: ${({ $flexDirection }) => $flexDirection || 'column'};

  && > *:not(${FieldSetExtrasContainer}) {
    margin-top: ${(props) => (props.$dimension === 's' ? 12 : 16)}px;
    margin-right: 16px;
  }

  ${Legend} {
    margin-top: 0;
  }
`;

export interface PropsType extends FieldsetHTMLAttributes<HTMLFieldSetElement> {
  legend?: ReactNode;
  dimension?: FieldSetDimension;

  /** Текст будет виден ниже компонента */
  extraText?: ReactNode;

  /** Установка статуса поля */
  status?: InputStatus;

  /** Установка статуса поле обязательно для заполнения */
  required?: boolean;

  /** Установка статуса disabled */
  disabled?: boolean;

  /** Расположение контента по вертикали или горизонтали */
  flexDirection?: 'column' | 'row';
}

export const FieldSet: FC<PropsType> = (props) => {
  const { legend, children, extraText, dimension, flexDirection, ...fieldsetProps } = props;
  return (
    <FieldSetContainer
      {...fieldsetProps}
      $dimension={dimension}
      $flexDirection={flexDirection}
      data-dimension={dimension}
      aria-required={props.required}
      data-status={props.status}
      data-required-within={props.required ? '' : undefined}
      data-disabled={props.disabled ? '' : undefined}
    >
      {legend && <Legend $dimension={dimension} children={legend} />}
      {children}
      {extraText && (
        <FieldSetExtrasContainer>
          <ExtraTextContainer>{extraText}</ExtraTextContainer>
        </FieldSetExtrasContainer>
      )}
    </FieldSetContainer>
  );
};

FieldSet.displayName = 'FieldSet';
