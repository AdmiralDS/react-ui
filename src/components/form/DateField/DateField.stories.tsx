import { DateField } from '#src/components/form/DateField';
import { INPUT_DIMENSIONS_VALUES } from '#src/components/input/types';
import * as React from 'react';
import type { ComponentMeta, ComponentStory } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import styled, { ThemeProvider } from 'styled-components';
import type { Theme } from '#src/components/themes';
import { DataAttributesDescription } from '#src/components/form/common';
import { ALL_BORDER_RADIUS_VALUES } from '#src/components/themes/borderRadius';

export default {
  title: 'Admiral-2.1/Form Field/DateField',
  component: DateField,
  decorators: [withDesign],
  parameters: {
    docs: {
      source: {
        code: null,
      },
    },
    componentSubtitle: <DataAttributesDescription />,
    design: [
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A53407',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A53678',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A53659',
      },
    ],
  },
  argTypes: {
    dimension: {
      options: INPUT_DIMENSIONS_VALUES,
      control: { type: 'radio' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    displayClearIcon: {
      control: { type: 'boolean' },
    },
    disableCopying: {
      control: { type: 'boolean' },
    },
    range: {
      control: { type: 'boolean' },
    },
    displayCharacterCounter: {
      control: { type: 'boolean' },
    },
    displayInline: {
      control: { type: 'boolean' },
    },
    defaultIsCalendarOpen: {
      control: { type: 'boolean' },
    },
    extraText: {
      control: { type: 'text' },
    },
    handleInput: {
      control: false,
    },
    placeholder: {
      control: { type: 'text' },
    },
    minDate: {
      control: false,
    },
    maxDate: {
      control: false,
    },
    themeBorderKind: {
      options: ALL_BORDER_RADIUS_VALUES,
      control: { type: 'radio' },
    },
    skeleton: {
      control: { type: 'boolean' },
    },
  },
} as ComponentMeta<typeof DateField>;

const FormContainer = styled.form`
  > * {
    margin-bottom: 24px;
  }
`;

const Template: ComponentStory<typeof DateField> = (props) => {
  const [localValue, setValue] = React.useState<string>('');

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.currentTarget.value;
    console.log(`handleChange ${inputValue}`);
    setValue(inputValue);
  };

  const handleFormBlur = (e: React.FocusEvent<HTMLFormElement>) => {
    console.log(`form blur:
    current target: ${e.currentTarget.id}
    target: ${e.target.id}
    related target: ${e.relatedTarget?.id}
    `);
  };

  const handleFormFocus = (e: React.FocusEvent<HTMLFormElement>) => {
    console.log(`form focus:
    current target: ${e.currentTarget.id}
    target: ${e.target.id}
    related target: ${e.relatedTarget?.id}
    `);
  };

  function swapBorder(theme: Theme): Theme {
    theme.shape.borderRadiusKind = (props as any).themeBorderKind || theme.shape.borderRadiusKind;
    return theme;
  }

  return (
    <ThemeProvider theme={swapBorder}>
      <FormContainer id="form 1" onBlur={handleFormBlur} onFocus={handleFormFocus}>
        <DateField
          data-container-id="dateFieldIdOne"
          {...props}
          value={localValue}
          onChange={handleChange}
          id={'date 1'}
          placeholder="Это placeholder"
        />
        <DateField data-container-id="dateFieldIdTwo" required label="uncontrolled input" id={'date 2'} />
        <DateField
          data-container-id="dateFieldIdThree"
          type="date-range"
          id="date range 1"
          label="uncontrolled date range"
          placeholder={'Введите отрезок времени'}
        />
      </FormContainer>
    </ThemeProvider>
  );
};

export const InputFieldInput = Template.bind({});

InputFieldInput.args = {
  label: 'Label',
};

InputFieldInput.storyName = 'DateField example';
