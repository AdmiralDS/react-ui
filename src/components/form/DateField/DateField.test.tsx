import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { LIGHT_THEME } from '#src/components/themes';
import { DateField } from '#src/components/form/DateField';
import { act } from 'react';

const testId = 'my test input';

describe('DateField', () => {
  test('should render component', () => {
    const { container } = render(
      <ThemeProvider theme={LIGHT_THEME}>
        <DateField id="some static snapshot id" data-testid={testId} />
      </ThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('should focus on component when user on press key tab', async () => {
    const user = userEvent.setup();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <DateField data-testid={testId} />
      </ThemeProvider>,
    );
    await act(async () => {
      await user.tab();
    });
    const input = screen.getByTestId(testId) as HTMLInputElement;
    await act(async () => {
      await user.keyboard('21.0');
    });
    expect(input.value).toBe('21.0_.____');
    await act(async () => {
      await user.keyboard('2.2023');
    });
    expect(input.value).toBe('21.02.2023');
  });
});
