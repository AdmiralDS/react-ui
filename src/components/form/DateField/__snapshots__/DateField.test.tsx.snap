// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField should render component 1`] = `
.c1 {
  position: relative;
  display: flex;
  align-items: stretch;
  border: none;
  border-radius: var(--admiral-border-radius-Medium, 4px);
  pointer-events: all;
}

.c2 {
  height: 40px;
}

.c0 {
  display: flex;
  overflow: hidden;
  flex-direction: column;
}

.c11 {
  display: flex;
  justify-content: space-between;
}

.c10 {
  cursor: pointer;
}

.c10 *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral50, #74767B);
}

[data-loading] .c10.c10.c10 {
  pointer-events: none;
}

[disabled] .c10.c10.c10 {
  pointer-events: none;
}

[disabled] .c10.c10.c10 *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral30, #B2B5B8);
}

.c10:hover *[fill^='#'] {
  fill: var(--admiral-color-Primary_Primary70, #0046E2);
}

.c7 {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  margin: 0;
  pointer-events: none;
  overflow: hidden;
  min-width: 0;
  background: none;
  border-radius: inherit;
  border: 1px solid var(--admiral-color-Neutral_Neutral40, #9EA0A4);
}

.c5 {
  outline: none;
  appearance: none;
  border-radius: inherit;
  box-sizing: border-box;
  flex: 1 1 auto;
  min-width: 10px;
  border: none;
  text-overflow: ellipsis;
  padding: 0 16px;
  font-family: var(--admiral-font-family, 'VTB Group UI', sans-serif);
  font-style: normal;
  font-weight: var(--admiral-font-weight-Body_Body1Long, 400);
  font-size: var(--admiral-font-size-Body_Body1Long, 16px);
  line-height: var(--admiral-line-height-Body_Body1Long, 24px);
  font-feature-settings: 'tnum' on,'lnum' on;
  text-rendering: geometricPrecision;
  color: var(--admiral-color-Neutral_Neutral90, #252629);
  background-color: var(--admiral-color-Neutral_Neutral00, #FFFFFF);
  padding-right: 48px;
  padding-left: 16px;
}

.c5.c5.c5:disabled {
  color: var(--admiral-color-Neutral_Neutral30, #B2B5B8);
}

.c5::placeholder {
  color: var(--admiral-color-Neutral_Neutral50, #74767B);
}

.c5:disabled::placeholder {
  color: var(--admiral-color-Neutral_Neutral30, #B2B5B8);
}

[data-disable-copying] .c5 {
  user-select: none;
}

[data-disable-copying] .c5::selection {
  background-color: transparent;
}

.c5.c5.c5:user-invalid+.c6 {
  border: 1px solid var(--admiral-color-Error_Error60Main, #D92020);
}

.c5.c5.c5:user-invalid:hover:not(:disabled)+.c6 {
  border: 1px solid var(--admiral-color-Error_Error70, #BA1717);
}

.c5.c5.c5:user-invalid:focus:not(:disabled)+.c6 {
  border: 2px solid var(--admiral-color-Error_Error60Main, #D92020);
}

[data-read-only] .c5.c5.c5,
.c5.c5.c5:disabled {
  background-color: var(--admiral-color-Neutral_Neutral10, #EAEBEC);
  border-color: transparent;
}

[data-loading] .c5.c5.c5,
.c5.c5.c5:disabled {
  cursor: not-allowed;
  pointer-events: none;
}

.c5 ::-ms-clear,
.c5 ::-ms-reveal {
  display: none;
}

.c8 {
  position: absolute;
  top: 0;
  bottom: 0;
  display: flex;
  align-items: center;
}

.c8 svg {
  border-radius: var(--admiral-border-radius-Medium, 4px);
  display: block;
  width: 24px;
}

.c8 svg:focus {
  outline: none;
}

.c8 svg:focus-visible {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

.c9 {
  right: 0;
  padding-right: 16px;
}

.c9>*:not(:first-child) {
  margin-left: 8px;
}

.c3:focus-within:not(:disabled)>.c6 {
  border: 2px solid var(--admiral-color-Primary_Primary60Main, #0062FF);
}

.c3:hover:not(:focus-within)>.c6 {
  border-color: var(--admiral-color-Neutral_Neutral60, #67696D);
}

.c4 {
  min-width: 150px;
}

<div>
  <div
    class="c0"
    data-field-id="some static snapshot id"
  >
    <div
      class=""
    >
      <div
        class="c1 c2 c3 c4"
      >
        <input
          class="c5"
          data-testid="my test input"
          id="some static snapshot id"
        />
        <div
          class="c6 c7"
        />
        <div
          class="c8 c9"
        >
          <div
            class="c10"
            tabindex="0"
          />
        </div>
      </div>
      <div
        class="c11"
      />
    </div>
  </div>
</div>
`;
